packages/tasit-action/src/ConfigLoader.test.js

```diff
@@ -1,40 +1,230 @@
+      etherscan = { ...etherscan, apiKey };
+    }
+
+    provider = { ...provider, etherscan };
+  } else if (providerType === "infura") {
+    if (!infura) infura = { apiKey: null };
+    else {
+      const { apiKey } = infura;
+      infura = { ...infura, apiKey };
+    }
+
+    provider = { ...provider, infura };
+  } else if (providerType === "jsonrpc") {
+    const { user, password, allowInsecure } = jsonRpc;
+    jsonRpc = { ...jsonRpc, user, password };
+
+    if (allowInsecure === undefined)
+      jsonRpc = { ...jsonRpc, allowInsecure: false };
+
+    provider = { ...provider, jsonRpc };
+  }
+
+  config = { ...config, provider };
+  return config;
+};
+
+const checkConfig = async config => {
+  const userConfig = fulfillWithDefaults(config);
+
+  ConfigLoader.setConfig(config);
+  const provider = ProviderFactory.getProvider();
+  const providerConfig = await extractProviderConfig(provider);
+  const eventsConfig = ConfigLoader.getConfig().events;
+  const loadedConfig = {
+    provider: providerConfig,
+    events: eventsConfig,
+  };
+
+  expect(userConfig).to.deep.equal(loadedConfig);
+};
 
-// Note: This test suite needs to be improved
 describe("TasitAction.ConfigLoader", () => {
   let defaultConfig;
 
-  before("save original config", async () => {
+  beforeEach("store default config", () => {
     defaultConfig = ConfigLoader.getConfig();
   });
 
-  after("back to original config", async () => {
+  afterEach("back to default config", () => {
     ConfigLoader.setConfig(defaultConfig);
   });
 
-  it("should config provider from ConfigLoader parameters", async () => {
+  it("should setup a provider connected to a local node using JsonRPC", async () => {
     const config = {
       provider: {
         network: "other",
         provider: "jsonrpc",
         pollingInterval: 50,
         jsonRpc: {
-          url: "http://some.node.eth",
-          port: 1234,
+          url: "http://localhost",
+          port: 8545,
         },
       },
       events: {
         timeout: 2000,
       },
     };
-    ConfigLoader.setConfig(config);
 
-    const provider = ProviderFactory.getProvider();
+    await checkConfig(config);
+  });
 
-    const { connection } = provider;
-    const { url } = connection;
+  it("should setup a provider connected to the rinkeby testnet using Fallback", async () => {
+    const config = {
+      provider: {
+        network: "rinkeby",
+        provider: "fallback",
+        pollingInterval: 4000,
+      },
+      events: {
+        timeout: 2000,
+      },
+    };
+
+    await checkConfig(config);
+  });
+
+  describe("should setup a provider connected to the ropsten testnet using Infura", async () => {
+    it("without apiKey", async () => {
+      const config = {
+        provider: {
+          network: "ropsten",
+          provider: "infura",
+          pollingInterval: 4000,
+        },
+        events: {
+          timeout: 2000,
+        },
+      };
+
+      await checkConfig(config);
```

Comment:
Do we need to worry about this test suite running too often and failing intermittently because of that?

Reply:
Sorry, I don't get it.

Reply:
Does our unit test suite that runs in CI hit infura? It looks like it. If so, is there a risk of being throttled if we open too many PRs, for instance?

As a general rule keeping unit tests local seems like a goal worth striving for

Reply:
But I can see why this would be a useful exception to that rule. Just food for thought, no action required before merging

Reply:
Hmmm I see. Yes, that test cases are hitting Infura and Etherscan. Maybe move both to integration tests in future?

Reply:
Yep, sounds good!